libraries/AP_UAVCAN/AP_UAVCAN.cpp

```diff
@@ -1645,4 +1690,51 @@ bool AP_UAVCAN::led_write(uint8_t led_index, uint8_t red, uint8_t green, uint8_t
     return true;
 }
 
+AP_UAVCAN::EscStatus_Data* AP_UAVCAN::escstatus_getptrto_data(uint8_t id)
+{
+    // I think the esc_index are continues, by how ArduPilot works
+    // so we could just directly jump with id into the data list, return &_escstatus.data[id], with id<8 overflow protection of course
+
+    // check if id is already in list, and if it is take it
+    for (uint8_t i = 0; i < AP_UAVCAN_ESCSTATUS_MAX_NUMBER; i++) {
+        if (_escstatus.id[i] == id) {
+            return &_escstatus.data[i];
+        }
+    }
+
+    // if id is not yet in list, find the first free spot, and take that
+    for (uint8_t i = 0; i < AP_UAVCAN_ESCSTATUS_MAX_NUMBER; i++) {
+        if (_escstatus.id[i] == UINT8_MAX) {
+            _escstatus.id[i] = id;
+            return &_escstatus.data[i];
+        }
+    }
+
+    return nullptr;
+}
+
+void AP_UAVCAN::escstatus_update_data(uint8_t id)
+{
+    // only 8 LOG_ESC1_MSG are defined, see /libraries/DataFlash/LogStructure.h
+    // technically, it could happen that the esc_index is not continues, and one would need a better handling
+    // however, I think, ArduPilot implicitly enforces continues esc_index, so should be no problem
+    if (id >= 8) return;
+
+    for (uint8_t i = 0; i < AP_UAVCAN_ESCSTATUS_MAX_NUMBER; i++) {
+        if (_escstatus.id[i] == id) {
+
+                    uint64_t time_us = AP_HAL::micros64();
+                    struct log_Esc pkt = {
+                            LOG_PACKET_HEADER_INIT((uint8_t)(LOG_ESC1_MSG + id)),
+                            time_us     : time_us,
+                            rpm         : (int16_t)(_escstatus.data[i].rpm),
+                            voltage     : (int16_t)(_escstatus.data[i].voltage*100.0f + 0.5f),
+                            current     : (int16_t)(_escstatus.data[i].current*100.0f + 0.5f),
+                            temperature : (int16_t)(_escstatus.data[i].temperature*100.0f + 0.5f)
+                     };
+                     DataFlash_Class::instance()->WriteBlock(&pkt, sizeof(pkt));
```

Comment:
I would prefer to make changes to libraries/DataFlash/LogFile.cpp to DataFlash_Class::Log_Write_ESC method to use logging that is already there. Otherwise we will have two places in code where logging of same parameter may occur.